chain/chain/src/chain.rs

```diff
@@ -2594,19 +2592,22 @@ impl Chain {
 
         let chunk = self.get_chunk_clone_from_header(chunk_header)?;
         let chunk_proof =
-            chunk_proofs.get(prev_shard_index).ok_or(Error::InvalidShardId(shard_id))?.clone();
-        let block_header =
+            chunk_proofs.get(shard_index).ok_or(Error::InvalidShardId(shard_id))?.clone();
+        let chunk_block_header =
             self.get_block_header_on_chain_by_height(&sync_hash, chunk_header.height_included())?;
 
         // Collecting the `prev` state.
         let (prev_chunk_header, prev_chunk_proof, prev_chunk_height_included) = match self
-            .get_block(block_header.prev_hash())
+            .get_block(chunk_block_header.prev_hash())
         {
             Ok(prev_block) => {
+                let (prev_shard_id, prev_shard_index) = self
+                    .epoch_manager
+                    .get_prev_shard_id(chunk_block_header.prev_hash(), shard_id)?;
```

Comment:
It took me a while to convince myself that this is indeed prev and not prev prev.

Reply:
You mean that `get_prev_shard_id` accepts `prev_hash`? Yeah... I'll go ahead and rename with `get_prev_shard_id_from_prev_hash`.

Reply:
Lol. Yeah, I agree.

We really really need to make this logic simpler and try to extract common logic. This block iteration and receipts aggregation is kinda duplicated in many places; and where we try to use common functions, there are usually too many parameters. I had some trouble with `get_state_witness_block_range`.
I also think it's worthy to show my thinking process.

First - I didn't know (forgot) that new sync hash guarantees prev_shard_id = shard_id, so I added this.
Second - I understood your comment, but thought that we still need to call get_prev_shard_id to support old sync hash.
Third - I realised that we are likely to enable CurrentEpochStateSync not later than SimpleNightshadeV4, so we don't need to call it.

But that's too much thinking which justifies the need for refactoring. Anyway, I guess yeah, the change makes logic slightly better.